### PR TITLE
[PIL-1033] - Prevent explicit exposure of clientPillar2Id within urls

### PR DESCRIPTION
**Pull request by jassalrichy**

### CODE REVIEW

File: app/controllers/IndexController.scala [hunk 1]
```diff
@@ -43,23 +47,27 @@ class IndexController @Inject() (
...
-      Redirect(routes.DashboardController.onPageLoad())
+      Redirect(routes.DashboardController.onPageLoad)
     } else {
       Redirect(routes.TaskListController.onPageLoad)
     }
 
   }
 
-  def onPageLoadBanner(clientPillar2Id: Option[String] = None): Action[AnyContent] = Action.async { implicit request =>
+  def onPageLoadBanner(): Action[AnyContent] = (bannerIdentity andThen getData andThen requireData).async { implicit request =>
```
Comment (TevynAllen) [inline app/controllers/IndexController.scala:57]: I'm not sure I understand the need for the `BannerIdentifierAction` considering this method already calls auth. maybe we can remove?

File: app/controllers/actions/BannerIdentifierAction.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import uk.gov.hmrc.auth.core.AuthProvider.GovernmentGateway
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait BannerIdentifierAction
```
Comment (TevynAllen) [inline app/controllers/actions/BannerIdentifierAction.scala:36]: Considering the usages. I don't think we need this action at all

File: app/controllers/actions/EnrolmentIdentifierAction.scala [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+import utils.Pillar2SessionKeys
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+trait EnrolmentIdentifierAction
+    extends ActionRefiner[Request, IdentifierRequest]
+    with ActionBuilder[IdentifierRequest, AnyContent]
+    with ActionFunction[Request, IdentifierRequest] {}
```
Comment (TevynAllen) [inline app/controllers/actions/EnrolmentIdentifierAction.scala:44]:
```suggestion
trait EnrolmentIdentifierAction
    extends ActionRefiner[Request, IdentifierRequest]
    with ActionBuilder[IdentifierRequest, AnyContent]
    with ActionFunction[Request, IdentifierRequest]
```

File: app/controllers/actions/EnrolmentIdentifierAction.scala [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+import utils.Pillar2SessionKeys
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+trait EnrolmentIdentifierAction
```
Comment (TevynAllen) [inline app/controllers/actions/EnrolmentIdentifierAction.scala:41]: Theres actually no difference between `EnrolmentIdentifierAction` and `IdentifierAction`. Can we just use `IdentifierAction` and use guice `Named` to distinguish between different instances 

Alternatively we could create different method that create instances of the `IdentifierAction`

File: app/controllers/actions/EnrolmentIdentifierAction.scala [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+          logger.info("EnrolmentAuthIdentifierAction - authAsAgent - Insufficient enrolment for Agent")
+          Future.successful(Left(Redirect(routes.AgentController.onPageLoadUnauthorised)))
+      }
+    }
+  }
+
+}
+
+@Singleton
+class ASAEnrolmentIdentifierAction @Inject() (
```
Comment (TevynAllen) [inline app/controllers/actions/EnrolmentIdentifierAction.scala:161]: Ditto 

Can this be moved to a separate file?

File: app/controllers/IndexController.scala (hunk 1 above)
Comment (jassalrichy) [reply app/controllers/IndexController.scala:57]: @TevynAllen 
In order to make a call to: request.userAnswers.get(AgentClientPillar2ReferencePage)
Only way I could get this to work was to create an action (bannerIdentifierAction) which returns the request.
Any suggestions on how to get around this ?

Comment (TevynAllen) [reply app/controllers/IndexController.scala:57]: maybe using the session repository 

```
  def onPageLoadBanner(): Action[AnyContent] = Action.async { implicit request =>
    authorised(AuthProviders(GovernmentGateway))
      .retrieve(
        Retrievals.internalId and Retrievals.allEnrolments
          and Retrievals.affinityGroup and Retrievals.credentialRole and Retrievals.credentials
      ) {
        case _ ~ e ~ Some(Organisation) ~ _ ~ _ if hasPillarEnrolment(e) =>
          Future successful Redirect(routes.DashboardController.onPageLoad)
        case _ ~ _ ~ Some(Organisation) ~ _ ~ _ => Future successful Redirect(routes.TaskListController.onPageLoad)
        case _ ~ _ ~ Some(Agent) ~ _ ~ _ =>
          OptionT(sessionRepository.get(AgentClientPillar2ReferencePage)).isDefined.map {
            case true  => Redirect(routes.DashboardController.onPageLoad)
            case false => Redirect(appConfig.asaHomePageUrl)
          }
        case _ ~ _ ~ Some(Individual) ~ _ ~ _ => Future successful Redirect(routes.UnauthorisedIndividualAffinityController.onPageLoad)
        case _                                => Future successful Redirect(controllers.routes.UnauthorisedController.onPageLoad)
      }
      .recover {
        case _: NoActiveSession =>
          Redirect(appConfig.loginUrl, Map("continue" -> Seq(appConfig.loginContinueUrl)))
        case _: AuthorisationException =>
          Redirect(controllers.routes.UnauthorisedController.onPageLoad)
      }
  }


```

Comment (jassalrichy) [reply app/controllers/IndexController.scala:57]: Perfect, Thnaks

File: app/controllers/actions/BannerIdentifierAction.scala (hunk 2 above)
Comment (jassalrichy) [reply app/controllers/actions/BannerIdentifierAction.scala:36]: Removed.

File: app/controllers/actions/EnrolmentIdentifierAction.scala (hunk 5 above)
Comment (jassalrichy) [reply app/controllers/actions/EnrolmentIdentifierAction.scala]: Done

File: app/controllers/actions/EnrolmentIdentifierAction.scala (hunk 4 above)
Comment (jassalrichy) [reply app/controllers/actions/EnrolmentIdentifierAction.scala]: Applied.

File: app/controllers/actions/EnrolmentIdentifierAction.scala (hunk 3 above)
Comment (jassalrichy) [reply app/controllers/actions/EnrolmentIdentifierAction.scala]: Done

**Review by TevynAllen — APPROVED** [review]

**Review by javed-iqbal1 — DISMISSED** [review]

**Review by javed-iqbal1 — APPROVED** [review]

**Review by AKKUM — APPROVED** [review]

**Pull request merged**